### ISSUE TITLE
Macros: Could `ClassTypeBuilder` have access to user-defined class methods?

### ISSUE THREAD

**Issue by rrousselGit** [dart-lang/language#3809]
Hello again!

It appears that there is no way for a `ClassTypeBuilder` to obtain the methods of the annotated class.  
This is particularly problematic in my case, because I need to add a mixin to the class that's based on the return value of a method present in said class.

Consider:

```dart
@riverpod
class Foo {
  Model build();
}
```

I want to generate:

```dart
import 'model.dart' as prefix0;

// New type:
class FooProvider extends Provider<prefix0.Model> {...}

// Added a mixin to the type that depend s
augment class Foo with Notifier<prefix0.Model> {
  ...
}
```

The problem is, both the newly declared type and the mixin applied on the existing type depends on the return value of that `build` method (and the output is not constant. There's some logic around it).
So I need to find that `build` method in a phase that's able to add new types and mixin. But there doesn't seem to be a way to do so, as far as I know.

For information, it is an error if users do not define that `build` method. And I don't care about cases where that `build` method is coming from a different macro or a mixins or a subclass. That `build` method should always be directly specified by users.

**Comment by rrousselGit** [comment]
This is critical for being able to migrate [Riverpod](https://pub.dev/packages/riverpod_generator) to macros.
Without a solution to this, I wouldn't be able to.


Maybe we could have `ClassDeclaration` list the user-defined methods ; while explicitly not supporting:
- inherited methods from extends/with/implements
- methods added through augmentation

**Comment by jakemac53** [comment]
I do think it is a critical part of the macros feature that macros can compose well - and that means supporting other macros generating the `build` method in this case. So, that is the general reasoning for why this is not supported today. You would have to require users to encode this type in the class header somehow, probably by writing `with Notifier<Model>` in this case.

That being said it should be possible to allow an API which only gives you the user defined methods, or maybe we would specify it more specifically as the pre-augmentation members (which would also mean not including explicit hand written augmentations). From a technical perspective that should be safe to do.

**Comment by rrousselGit** [comment]
Thinking about it, Freezed would have the same issue.
Because it needs to generate 2-3 classes per constructor. So we'd need to have the list of constructors in the type phase ; which we can't yet.

**Comment by rrousselGit** [comment]
Alternatively, what is the reason why the "declaration" phase can't emit classes?
It can output variables, so I don't get why it can't output classes.

**Comment by jakemac53** [comment]
> Alternatively, what is the reason why the "declaration" phase can't emit classes?

Because it is allowed to resolve types to their declarations. If macros in this phase could emit new types, it could change how types resolve (maybe changing an error to a success, or maybe a success to a totally different declaration in the case of shadowing).

We try hard to eliminate the cases where a macro might ask a question and get a wrong (or inconsistent) answer. Sometimes, the answer can be incomplete (in the declarations phase), but this is the one exception today.

**Comment by jakemac53** [comment]
Are the generated types intended to be user visible? Or would they always be private to this library, and in particular private to code generated by this macro? We have considered, and I think maybe the proposal even mentions, the ability to emit macro-private code in later phases, that otherwise wouldn't be allowed.

**Comment by rrousselGit** [comment]
>  If macros in this phase could emit new types, it could change how types resolve (maybe changing an error to a success, or maybe a success to a totally different declaration in the case of shadowing).

That's never going to happen for my scenario.

I want to generate classes that are implementation details of the macro. They are your typical `_$Foo`.

**Comment by rrousselGit** [comment]
Same thing for the annotated class. I want to add mixins on it, but that class is never going to be instantiated by users.

Given:
```dart
@macro
class Foo {}
```

It is an error if any constructor is defined. And a `Foo._();` constructor will be generated, on top of the `with Notifier<T>` added.

**Comment by rrousselGit** [comment]
(I deleted my previous message, I think that was incorrect).

My classes _may_ be public. But they aren't used by others. I'm supposed to be the sole consumer of those classes, but they are technically visible.
This is because I want to generate a variable that's a public instance of those classes.

Given:
```dart
@riverpod
class Foo {
   int build();
}
```

We have:

```dart
const fooProvider = FooProvider<int>._(Foo.new);

class FooProvider extends Provider<int> {
  // Only my macro can instantiate this new class, but it is public.
  const FooProvider._();
}

augment class Foo with Notifier<int> {
  // Only my macro is allowed to instantiate the annotated class.
  Foo._();
}
```

**Comment by rrousselGit** [comment]
Fun fact: My macro can also be applied on functions, like so:

```dart
@riverpod
int foo() {

}
```

But in this scenario, I have no issue. Because `FunctionDeclaration.returnType` is available.

**Comment by rrousselGit** [comment]
I'm thing I'm currently looking into is doing:

```dart
@Riverpod<int>()
class Foo {
  int build();
}
```

Although I don't see a way to convert that generic into an Identifier. 

And I quite dislike how this involves duplicating the type information.

**Comment by jakemac53** [comment]
Yes generic type parameters won't work for macros because we can't take the types from the users program and materialize them in the macro program.

The intention is to pass these as regular arguments, and have the coerced into a TypeAnnotation implicitly, but that isn't implemented yet.

**Comment by rrousselGit** [comment]
I see thanks.

With regards to removing the code duplication: Would it be possible to have users declare `build` without a return type, and have the augmentation add it?

```dart
@Riverpod(int)
class Foo {
   build() => ...;
}

augment class Foo {
  augment int build();
}
```

**Comment by jakemac53** [comment]
> With regards to removing the code duplication: Would it be possible to have users declare `build` without a return type, and have the augmentation add it?

We don't allow augmentations to change the type of something pre-existing, so no we wouldn't allow this. Same reason we don't allow adding optional parameters. I get why that is limiting, but it is an intentional design choice for readability/usability.

If all these classes are implementing a `T build()` method you could consider an interface which is explicit (and maybe you still have subtypes which are more specific but implement that?).

```
@Riverpod()
class Foo implements Thing<int> {
  build() => ...; // Gets the inferred return type `int`.
}

abstract interface class Thing<T> {
  T build();
}
```

**Comment by rrousselGit** [comment]
> If all these classes are implementing a T build() method you could consider an interface which is explicit (and maybe you still have subtypes which are more specific but implement that?).

No. The prototype of `build` isn't fixed.

Users can freely add arguments ; which the code-generator deals with.

In particular, the following is valid too:

```dart
@Riverpod();
class Foo {
  Future<int> build(int a, {required int named, int optional = 42]) {...}
}
```

Which then generated an object used as:

```dart
ref.watch(fooProvider(42, named: 42));
```

So there's no shared interface.

**Comment by rrousselGit** [comment]
> We don't allow augmentations to change the type of something pre-existing, so no we wouldn't allow this. 

Although in this case, we're not replacing something existing. We're specifying something that wasn't specified yet.

Kind of like how we support:

```dart
@macro
int fn();
```

Then followed by an augmentation that adds a body.

(... although this is unimplemented. I assume this will be supported, right?)

**Comment by jakemac53** [comment]
Omitting a type annotation on a function means it is inferred - and augmentations are not allowed to affect inference (by removing it through making the type explicit or altering the inferred type).

The only actual inference that can take place for return types of methods afaik is copying the type from the super declaration if one exists, otherwise you are getting `dynamic`.

In other words, while a user did not write a type there, that does not mean there is no type. They just didn't declare the type explicitly.

Similarly, augmentations cannot add types to parameters where one was not declared initially, etc.

**Comment by rrousselGit** [comment]
Sounds like you're saying that there's no easy way forward.

Do you think it is realistic to expect `ClassDeclaration` to expose user-defined methods, even at the cost of composition in this case?

**Comment by jakemac53** [comment]
I don't think we want to put things directly on the `ClassDeclaration` instance. But it is possible we could allow asking for members in the types phase, with the understanding that they will not be complete (which is already the case in the declarations phase anyways, to a lesser extent).

**Comment by rrousselGit** [comment]
Awesome, thanks!
I really hope that this makes it through. Otherwise it'd be quite hard for my users to upgrade to macros, and the syntax would be worse overall.

**Comment by StarProxima** [comment]
> to put things directly on the `ClassDeclaration` instance. But it is possible we could allow asking for members in the types phase, with the understanding that they will not be complete (which is already the case in the declarations phase anyways, to a lesser extent).

I was also experimenting with macros and came across this limitation.

The ability to get constructors, methods and fields, even if only user-defined, in the type phase seems pretty important. I think this is necessary to migrate users of some packages from build_runner to the macro system without pain
